build/patches/Enable-crash-reporting-ui.patch

```diff
@@ -0,0 +1,206 @@
+ import java.util.zip.GZIPOutputStream;
+ 
++import org.chromium.base.PathUtils;
++
+ /**
+  * This class tries to upload a minidump to the crash server.
+  *
+@@ -120,30 +124,18 @@ public class MinidumpUploader {
+             if (fileToUpload == null || !fileToUpload.exists()) {
+                 return Result.failure("Crash report does not exist");
+             }
+-            HttpURLConnection connection =
+-                    mHttpURLConnectionFactory.createHttpURLConnection(CRASH_URL_STRING);
+-            if (connection == null) {
+-                return Result.failure("Failed to create connection");
+-            }
+-            configureConnectionForHttpPost(connection, readBoundary(fileToUpload));
+-
+-            try (InputStream minidumpInputStream = new FileInputStream(fileToUpload);
+-                    OutputStream requestBodyStream =
+-                            new GZIPOutputStream(connection.getOutputStream())) {
+-                streamCopy(minidumpInputStream, requestBodyStream);
+-                int responseCode = connection.getResponseCode();
+-                if (isSuccessful(responseCode)) {
+-                    // The crash server returns the crash ID in the response body.
+-                    String responseContent = getResponseContentAsString(connection);
+-                    String uploadId = responseContent != null ? responseContent : "unknown";
+-                    return Result.success(uploadId);
+-                } else {
+-                    // Return the remote error code and message.
+-                    return Result.uploadError(responseCode, connection.getResponseMessage());
+-                }
+-            } finally {
+-                connection.disconnect();
+-            }
++
++            File outputDir = new File(PathUtils.getDownloadsDirectory());
++            File destination = new File(outputDir, fileToUpload.getName());
++
++            InputStream minidumpInputStream = new FileInputStream(fileToUpload);
++            OutputStream out = new BufferedOutputStream(
++                    new FileOutputStream(destination));
++            OutputStream requestBodyStream =
++                    new GZIPOutputStream(out);
++
++            streamCopy(minidumpInputStream, requestBodyStream);
++            return Result.success(destination.getPath());
+         } catch (IOException | RuntimeException e) {
+             return Result.failure(e.toString());
+         }
+diff --git a/components/upload_list/upload_list.cc b/components/upload_list/upload_list.cc
+--- a/components/upload_list/upload_list.cc
++++ b/components/upload_list/upload_list.cc
+@@ -20,7 +20,7 @@ namespace {
+ // chrome://webrtc-logs and the feedback UI. See https://crbug.com/972526.
+ constexpr base::TaskTraits kLoadingTaskTraits = {
+     base::MayBlock(), base::TaskPriority::USER_BLOCKING,
+-    base::TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN};
++    base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN};
```

Comment:
this is bad, and needs to be reviewed, although not particularly necessary right away: it happens that without it, a dcheck fails because somewhere it requires a singleton (not initialized) that it cannot generate in a (background) thread (the one IO that retrieves the crash list on file) because it would be temporary and _could_ lead to unwanted crashes.
I tried to look for it either, but I have to recompile everything with special flags that allow me to know the type of a template at runtime (the class is in fact a Lazy<T> and I need T).
But as I said, nothing happens and it does not crash, simply chromium is telling us that this is not good, but it works, let's say, since it's being used a lot, the benefits outweigh the risks

Reply:
this is the stack of that dcheck, really impossible to tell where it came from

```
Symbolizing stack using ABI: x86
Build fingerprint: 'Android/sdk_phone_x86_64/generic_x86_64:9/PSR1.180720.012/4923214:userdebug/test-keys'
Revision: '0'
pid: 13049, tid: 13117, name: ThreadPoolForeg  >>> org.chromium.chrome.stable <<<
signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
[FATAL:thread_restrictions.cc(207)] Check failed: !g_singleton_disallowed.Get().Get(). LazyInstance/Singleton is not allowed to be used on this thread. Most likely it's because this thread is not joinable (or the current task is running>
Abort message: '[FATAL:thread_restrictions.cc(207)] Check failed: !g_singleton_disallowed.Get().Get(). LazyInstance/Singleton is not allowed to be used on this thread. Most likely it's because this thread is not joinable (or the current>

Stack Trace:
  RELADDR   FUNCTION                   FILE:LINE
  00000b59  __kernel_vsyscall+9        /home/uazo/chromium/src/[vdso:eb55e000]
  0001fc08  syscall+40                 /system/lib/libc.so
  000321f3  abort+115                  /system/lib/libc.so
  07428a6a  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  0737fe1f  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  07380134  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  07367703  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  074185d3  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  076d2f6b  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  076d307b  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  076f9943  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  076f9857  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  076f9a80  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  072a60ef  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  072a5ea1  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  09770a1c  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  097709cc  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  09770981  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  09771217  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  09770f36  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  097710b3  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  09771079  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  0977104f  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  0377d545  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  074109b8  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  07410cfc  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  0377d545  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  073d43c0  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  07400445  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  074000d9  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  073ffc44  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  074366c1  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  073ff687  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  07409895  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  0740962f  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  0740951b  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  07436f29  offset 0x1b94000           /data/app/org.chromium.chrome.stable-6fQWTqIq06SqUKHnkvEPnQ==/libchrome.so
  0009cce5  __pthread_start(void*)+53  /system/lib/libc.so
  00033c1b  __start_thread+75          /system/lib/libc.so
```

Reply:
> this is the stack of that dcheck, really impossible to tell where it came from

Then why not initialize this when the UI starts?

Reply:
> Then why not initialize this when the UI starts?

I think it is something related to upload, like a service or some component that is under `IsMetricsAndCrashReportingEnabled` that remains false and therefore does not instantiated at startup, but I don't understand what, the code is a bit complex to read (a bit tangled)

Reply:
Could it be due to some other patch in Bromite? There are at least 2 patches touching the code related to uploading crash reports.

Reply:
> Could it be due to some other patch in Bromite?

I do not think so.
i have to try to change the clang flags, now it's more complex because i use goma, but i try.

Reply:
i found the problem, i think chrome is also affected.

the failed dcheck `AssertSingletonAllowed` is the fault of `ui::FormatBytes()` https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/crash_upload_list/crash_upload_list_android.cc;l=140;drc=221e331b49dfefadbc6fa40b0c68e6f97606d0b3;bpv=1;bpt=1

because it ultimately goes to use the icu library as a singleton.
https://source.chromium.org/chromium/chromium/src/+/master:base/i18n/number_formatting.cc;l=68;drc=221e331b49dfefadbc6fa40b0c68e6f97606d0b3;bpv=1;bpt=1

I preferred to delete that call, also because the file_size field is not used anywhere in the code. the alternative is to initialize the library before the thread fork, indeed, it seems strange to me that it is not already .

---
I write it here maybe I need it other times, to retrieve the name of T without changing clang flags:
```
  template <typename T> const char *name()
  {
      return __PRETTY_FUNCTION__;
  }

```